src/imitation/algorithms/pebble/entropy_reward.py

```diff
@@ -0,0 +1,125 @@
+"""Reward function for the PEBBLE training algorithm."""
+
+from enum import Enum, auto
+from typing import Dict, Optional, Tuple, Union
+
+import numpy as np
+import torch as th
+
+from imitation.policies.replay_buffer_wrapper import (
+    ReplayBufferRewardWrapper,
+    ReplayBufferView,
+    ReplayBufferAwareRewardFn,
+)
+from imitation.rewards.reward_function import RewardFn
+from imitation.util import util
+from imitation.util.networks import RunningNorm
+
+
+class PebbleRewardPhase(Enum):
+    """States representing different behaviors for PebbleStateEntropyReward."""
+
+    UNSUPERVISED_EXPLORATION = auto()  # Entropy based reward
+    POLICY_AND_REWARD_LEARNING = auto()  # Learned reward
+
+
+class PebbleStateEntropyReward(ReplayBufferAwareRewardFn):
+    """Reward function for implementation of the PEBBLE learning algorithm.
+
+    See https://arxiv.org/abs/2106.05091 .
+
+    The rewards returned by this function go through the three phases:
+    1. Before enough samples are collected for entropy calculation, the
+        underlying function is returned. This shouldn't matter because
+        OffPolicyAlgorithms have an initialization period for `learning_starts`
+        timesteps.
+    2. During the unsupervised exploration phase, entropy based reward is returned
+    3. After unsupervised exploration phase is finished, the underlying learned
+        reward is returned.
+
+    The second phase requires that a buffer with observations to compare against is
+    supplied with set_replay_buffer() or on_replay_buffer_initialized().
+    To transition to the last phase, unsupervised_exploration_finish() needs
+    to be called.
+    """
+
+    def __init__(
+        self,
+        learned_reward_fn: RewardFn,
+        nearest_neighbor_k: int = 5,
+    ):
+        """Builds this class.
+
+        Args:
+            learned_reward_fn: The learned reward function used after unsupervised
+                exploration is finished
+            nearest_neighbor_k: Parameter for entropy computation (see
+                compute_state_entropy())
+        """
+        self.learned_reward_fn = learned_reward_fn
+        self.nearest_neighbor_k = nearest_neighbor_k
+        self.entropy_stats = RunningNorm(1)
+        self.state = PebbleRewardPhase.UNSUPERVISED_EXPLORATION
+
+        # These two need to be set with set_replay_buffer():
+        self.replay_buffer_view: Optional[ReplayBufferView] = None
+        self.obs_shape: Union[Tuple[int, ...], Dict[str, Tuple[int, ...]], None] = None
+
+    def on_replay_buffer_initialized(self, replay_buffer: ReplayBufferRewardWrapper):
```

Comment:
When designing the API of `ReplayBufferAwareRewardFn`, it seemed more natural and decoupled to have a signature of 

    on_replay_buffer_initialized(self, replay_buffer: ReplayBufferRewardWrapper)

rather than

    on_replay_buffer_initialized(self, replay_buffer_view: ReplayBufferView, obs_shape: Union[Tuple[int, ...], Dict[str, Tuple[int, ...]]])

since potential future replay buffer aware functions may need something other than `obs_shape` as well. It's a minor reason, though.

Reply:
It made sense in tests, but I used a different solution and removed `set_replay_buffer()` as suggested.